File: 0029 Divide Two Integers/solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int divide(int dividend, int divisor) {
        if (divisor == 0)  return INT_MAX;
        if (dividend == INT_MIN && divisor == -1)  return INT_MAX;
        
        long long m = abs((long)dividend), n = abs((long)divisor);
        long long res = 0;
        
        while (m >= n)
        {
            long long t = n, shift = 1;
            while (2 * t <= m)
            {
                t *= 2;
                shift *= 2;
            }
            res += shift;
            m -= t;
        }
        return (dividend > 0) ^ (divisor > 0) ? -res : res;
    }
};

int main(){
    assert(Solution().divide(10, 3) == 3);
    return 0;
}

File: 0029 Divide Two Integers/solution.py
#! python3

class Solution:
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        flag = 1 if (dividend<0 and divisor<0) or (dividend>0 and divisor>0)  else -1
        dividend, divisor = abs(dividend), abs(divisor)
        res = 0
        while dividend >= divisor:
            tmp, i = divisor, 1
            while dividend >= tmp:
                dividend -= tmp
                res += i
                tmp <<= 1
                i   <<= 1
        if flag < 0:
            res = -res
            
        return min(max(-2147483648, res), 2147483647)

if __name__ == "__main__":
    print(Solution().divide(10, 3))